File: myFinance_improved_V2/lib/features/sales_invoice/domain/entities/invoice.dart
import 'package:equatable/equatable.dart';
import 'customer.dart';
import 'invoice_amounts.dart';
import 'items_summary.dart';

/// Cash location info for invoice
class InvoiceCashLocation extends Equatable {
  final String cashLocationId;
  final String locationName;
  final String locationType; // 'cash', 'bank', 'vault'

  const InvoiceCashLocation({
    required this.cashLocationId,
    required this.locationName,
    required this.locationType,
  });

  @override
  List<Object?> get props => [cashLocationId, locationName, locationType];
}

/// Core Invoice entity representing a sales invoice
class Invoice extends Equatable {
  final String invoiceId;
  final String invoiceNumber;
  final DateTime saleDate;
  final String status;
  final Customer? customer;
  final String storeId;
  final String storeName;
  final String storeCode;
  final InvoiceCashLocation? cashLocation;
  final String paymentMethod;
  final String paymentStatus;
  final InvoiceAmounts amounts;
  final ItemsSummary itemsSummary;
  final String? aiDescription;
  final String? createdById;
  final String? createdByName;
  final String? createdByEmail;
  final DateTime createdAt;

  const Invoice({
    required this.invoiceId,
    required this.invoiceNumber,
    required this.saleDate,
    required this.status,
    this.customer,
    required this.storeId,
    required this.storeName,
    required this.storeCode,
    this.cashLocation,
    required this.paymentMethod,
    required this.paymentStatus,
    required this.amounts,
    required this.itemsSummary,
    this.aiDescription,
    this.createdById,
    this.createdByName,
    this.createdByEmail,
    required this.createdAt,
  });

  /// Helper getters for status checking
  bool get isCompleted => status == 'completed';
  bool get isDraft => status == 'draft';
  bool get isCancelled => status == 'cancelled';

  /// Helper getters for payment status
  bool get isPaid => paymentStatus == 'paid';
  bool get isPending => paymentStatus == 'pending';

  // NOTE: timeString and dateString moved to presentation/extensions/invoice_display_extension.dart
  // UI formatting is a Presentation concern, not Domain

  @override
  List<Object?> get props => [
        invoiceId,
        invoiceNumber,
        saleDate,
        status,
        customer,
        storeId,
        storeName,
        storeCode,
        cashLocation,
        paymentMethod,
        paymentStatus,
        amounts,
        itemsSummary,
        aiDescription,
        createdById,
        createdByName,
        createdByEmail,
        createdAt,
      ];
}

File: website/src/features/invoice/domain/entities/Invoice.ts
/**
 * Invoice Entity
 * Represents a sales invoice with customer and item details
 */

import { DateTimeUtils } from '@/core/utils/datetime-utils';

export class Invoice {
  constructor(
    public readonly invoiceId: string,
    public readonly invoiceNumber: string,
    public readonly invoiceDate: string,
    public readonly storeId: string,
    public readonly cashLocationId: string | null,
    public readonly customerName: string,
    public readonly customerPhone: string | null,
    public readonly itemCount: number,
    public readonly totalQuantity: number,
    public readonly totalAmount: number,
    public readonly status: 'draft' | 'issued' | 'paid' | 'cancelled' | 'completed',
    public readonly currencySymbol: string,
    public readonly paymentMethod: 'cash' | 'card' | 'bank' | 'transfer' | string = 'cash',
    public readonly paymentStatus: 'paid' | 'pending' | 'cancelled' | string = 'pending',
    public readonly totalCost: number = 0,
    public readonly profit: number = 0,
    // Store info
    public readonly storeName: string = '',
    public readonly storeCode: string = '',
    // Cash location info
    public readonly cashLocationName: string | null = null,
    public readonly cashLocationType: string | null = null,
    // Amount details
    public readonly subtotal: number = 0,
    public readonly taxAmount: number = 0,
    public readonly discountAmount: number = 0,
    // Created by info
    public readonly createdByName: string = '',
    public readonly createdByEmail: string = '',
    // Created at timestamp
    public readonly createdAt: string = ''
  ) {}

  /**
   * Get status display
   */
  get statusDisplay(): string {
    switch (this.status) {
      case 'draft':
        return 'Draft';
      case 'issued':
        return 'Issued';
      case 'paid':
        return 'Paid';
      case 'completed':
        return 'Completed';
      case 'cancelled':
        return 'Cancelled';
      default:
        return this.status;
    }
  }

  /**
   * Format currency amount
   */
  formatCurrency(amount: number | null | undefined): string {
    const safeAmount = amount ?? 0;
    return `${this.currencySymbol}${safeAmount.toLocaleString('en-US', {
      minimumFractionDigits: 0,
      maximumFractionDigits: 0,
    })}`;
  }

  /**
   * Get formatted date
   * Note: invoiceDate is already converted to local timezone in InvoiceModel
   */
  get formattedDate(): string {
    if (!this.invoiceDate) {
      return 'Invalid Date';
    }

    try {
      // invoiceDate is already in local timezone, just parse and format
      const localDate = new Date(this.invoiceDate);

      // Format using Korean locale: "2025. 10. 25."
      return localDate.toLocaleDateString('ko-KR', {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
      });
    } catch (error) {
      console.error('Error formatting date:', this.invoiceDate, error);
      return 'Invalid Date';
    }
  }

  /**
   * Check if invoice is editable
   */
  get isEditable(): boolean {
    return this.status === 'draft';
  }

  /**
   * Check if invoice can be cancelled
   */
  get isCancellable(): boolean {
    return this.status === 'draft' || this.status === 'issued';
  }

  /**
   * Get payment method display text
   */
  get paymentMethodDisplay(): string {
    const methodMap: Record<string, string> = {
      'cash': 'Cash',
      'card': 'Card',
      'bank': 'Bank',
      'transfer': 'Bank',
      'bank_transfer': 'Bank',
    };
    return methodMap[this.paymentMethod] || this.paymentMethod;
  }

  /**
   * Get payment badge class for styling based on payment method
   */
  get paymentBadgeClass(): string {
    const method = this.paymentMethod.toLowerCase();
    if (method === 'bank' || method === 'transfer') {
      return 'payment-bank';
    } else if (method === 'card') {
      return 'payment-card';
    } else if (method === 'cash') {
      return 'payment-cash';
    }
    return 'payment-default';
  }

  /**
   * Get formatted time (HH:MM:SS)
   * Extracts time from createdAt timestamp
   */
  get formattedTime(): string {
    if (!this.createdAt) {
      return '';
    }

    try {
      const date = new Date(this.createdAt);
      return date.toLocaleTimeString('en-US', {
        hour: '2-digit',
        minute: '2-digit',
        second: '2-digit',
        hour12: false,
      });
    } catch (error) {
      console.error('Error formatting time:', this.createdAt, error);
      return '';
    }
  }
}
